django/utils/timesince.py

```diff
@@ -69,22 +71,38 @@ def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
             leapdays += 1
     delta -= datetime.timedelta(leapdays)
 
+    # fix months duration to be relative to the months between d and now
+    months = now.month - d.month
+    if months > 0:
+        m_days = (sum(MONTHS_DAYS[d.month - 1 : now.month - 1])) / months
+    else:
+        m_days = (365 - sum(MONTHS_DAYS[now.month - 1 : d.month - 1])) / (12 + months)
+    timesince_chunks_adjusted = list(TIMESINCE_CHUNKS)
+    timesince_chunks_adjusted[1] = (int(m_days * 24 * 60 * 60), "month")
+
     # ignore microseconds
     since = delta.days * 24 * 60 * 60 + delta.seconds
     if since <= 0:
         # d is in the future compared to now, stop processing.
         return avoid_wrapping(time_strings["minute"] % {"num": 0})
-    for i, (seconds, name) in enumerate(TIMESINCE_CHUNKS):
+    for i, (seconds, name) in enumerate(timesince_chunks_adjusted):
         count = since // seconds
         if count != 0:
             break
     else:
         return avoid_wrapping(time_strings["minute"] % {"num": 0})
+
     result = []
     current_depth = 0
-    while i < len(TIMESINCE_CHUNKS) and current_depth < depth:
-        seconds, name = TIMESINCE_CHUNKS[i]
+
+    if months <= 0:
+        months = 12 + months
+
+    while i < len(timesince_chunks_adjusted) and current_depth < depth:
+        seconds, name = timesince_chunks_adjusted[i]
         count = since // seconds
+        if name == "month":
+            count = min(count, months)
```

Comment:
I need to have a think about this. If the timesince chucks are correct, why do we still need this? 🤔 (Is there an example?)

Reply:
This was quite tricky and I discover it only when I wrote all the tests:
Consider this example:

- `d = 2022/1/1`
- `now = 2022/5/31`

let's see what the algorithm does:

1. months is 5-1 = 4
2. the average month length in this time frame is (31+28+31+30) / 4 = 30 days

if we look at the dates, between d and now there are exactly 150 days...
and if the month duration is 30 the result of `since // seconds` will be 5 instead of 4.

Without that adjustment, the algorithm might fail in cases where the average month duration between `d` and `now`
is smaller than the month duration of the month for `now`.

There are still other cases where, even with the adjustment, the result is questionable, consider this example:

- `d = 2022/1/31`
- `now = 2022/3/1`

returns `4 weeks, 1 day`...but an entire month has already passed, so `1 month` would be the correct result.
I would argue that `4 weeks, 1 day` is also correct, it is exactly 29 days, the days elapsed in the timeframe.

In conclusion, the current implementation is not yet 100% bulletproof, but it is an improvement over the previous 30 days constant.

Reply:
OK, thanks for this. 

I'm going to say we can leave the February case.